feat: api_server and client supporting list_custom_data

arenstar opened the pull request:
<!-- Thank you for sending your pull request. But first, have you included
unit tests, and is your code PEP8 conformant? [More details](https://github.com/freqtrade/freqtrade/blob/develop/CONTRIBUTING.md)
-->
## Summary

Wanted to bring support to the API server/RPC for accessing the trades custom_data.

Solve the issue: # No issue

## Quick changelog

- Added ft_client endpoint for list_custom_data
- Added schema ListCustomData 
- Added route for api_server_v1 /list_custom_data
- Updated _rpc_list_custom_data() to always return a list

arenstar committed:
feat: api_server and client supporting list_custom_data

xmatthias commented on freqtrade/rpc/rpc.py:
```diff
@@ -1097,19 +1097,28 @@ def _rpc_delete(self, trade_id: int) -> dict[str, str | int]:
+            trades = Trade.get_trades(trade_filter=[Trade.id == trade_id]).all()
+        else:
+            # If no trade_id, get all trades
+            trades = Trade.get_trades().all()
```
I'd not do this if i'm honest. 
this can yield responses for 1000nds of trades - creating an insanely large response.

if necessary - go with "all open trades" instead?

arenstar replied:
Ok so ive updated this to use Trade.get_open_trades() ^^

xmatthias reviewed (commented):
I'd group this in the "trades" section (`/trades/ .../`) - it's data related to the trade - having it "fully outside" makes it kinda look like it's something standalone.

also, if it's for "one" trade only (current interface) - you might wanna also use `/trades/{tradeid}/custom-data` ...

If we use get parameters (current interface would use `/list_custom_data?trade_id=22`) - then i think the endpoint should support passing multiple trades (and then return a list nested one level deeper).

---

On the other hand - i'd not support "all" trades.
the only thing making sense in an "all" context in my opinion is an "all open trades" - so maybe something like this could work - splitting the endpoint in two so the response is actually clear.

`/trades/<tradeid>/custom-data`
`/trades/all-open/custom-data`

I also think for the "list" - you'll want to return a somewhat nested list.
1 trade can have anything between 0 and x custom data endpoints.
if we return this as a flat list - we move responsibility to "assign to trade" to the consumer.

a response along the following lines might make using the data easier, as you get a (top level) list  of length "trade" back.
it'll also make identifying "this trade had no custom-data" easier - otherwise it's a missing entry in a list.

```
[
  {
    "trade_id": 1,
    "custom_data": [
{...}
}
]
```

Bambam-jow commented on freqtrade/rpc/rpc.py:
```diff
@@ -1097,19 +1097,28 @@ def _rpc_delete(self, trade_id: int) -> dict[str, str | int]:
 
-    def _rpc_list_custom_data(self, trade_id: int, key: str | None) -> list[dict[str, Any]]:
-        # Query for trade
-        trade = Trade.get_trades(trade_filter=[Trade.id == trade_id]).first()
```
Fgdxn

arenstar replied:
@Bambam-jow please help me understand your comment

xmatthias replied:
@arenstar best ignore - it's an account created yesterday - with about a 99% chance of being a scam account.
We get these from time to time - there's little we can do that won't have negative side-effects (like - limit to existing contributors, or prevent new accounts from interacting).

arenstar committed:
fix: no tradeID only returns open trades

arenstar committed:
fix: api url and rm key

arenstar committed:
fix

arenstar commented:
@xmatthias 
ive now updated the api_server endpoint to 
`/trades/<tradeid>/custom_data`

i also removed the key part as i just dont see why you would require that

arenstar commented:
Regarding the format of the response.

I thought a list regardless of it being a single trade or many is a common format that makes sense for an API.

my current testing is show this 

```
$ freqtrade-client -c user_data/config.json  list_custom_data

{"custom_data": [{"id": 3, "ft_trade_id": 2, "cd_key": "test1", "cd_type": "str", "cd_value": "1", "created_at": "2024-12-16T01:27:45.143719", "updated_at": "2024-12-16T10:12:30.125379"}, {"id": 4, "ft_trade_id": 2, "cd_key": "test2", "cd_type": "str", "cd_value": "2", "created_at": "2024-12-16T01:27:45.144628", "updated_at": "2024-12-16T10:12:30.126830"}]}
```

-------

> a response along the following lines might make using the data easier, as you get a (top level) list of length "trade" back. it'll also make identifying "this trade had no custom-data" easier - otherwise it's a missing entry in a list.


Ahh right, i see your point.
Ill need to adjust the structure and push another commit

arenstar committed:
fix: returned object

arenstar commented:
@xmatthias ok so now it returns an object like this

[{"trade_id": 2, "custom_data": [{"id": 3, "ft_trade_id": 2, "cd_key": "test1", "cd_type": "str", "cd_value": "1", "created_at": "2024-12-16T01:27:45.143719", "updated_at": "2024-12-16T10:12:30.125379"}, {"id": 4, "ft_trade_id": 2, "cd_key": "test2", "cd_type": "str", "cd_value": "2", "created_at": "2024-12-16T01:27:45.144628", "updated_at": "2024-12-16T10:12:30.126830"}]}]

xmatthias commented on freqtrade/rpc/rpc.py:
```diff
@@ -1097,19 +1097,23 @@ def _rpc_delete(self, trade_id: int) -> dict[str, str | int]:
-        # Query for trade
-        trade = Trade.get_trades(trade_filter=[Trade.id == trade_id]).first()
-        if trade is None:
+    def _rpc_list_custom_data(self, trade_id: int) -> list[dict[str, Any]]:
```
well won't this change (removal of the key argument) break how the telegram integration works? (written before CI has ran)

xmatthias commented on freqtrade/rpc/rpc.py:
```diff
@@ -1097,19 +1097,23 @@ def _rpc_delete(self, trade_id: int) -> dict[str, str | int]:
-        if trade is None:
+    def _rpc_list_custom_data(self, trade_id: int) -> list[dict[str, Any]]:
+        # Query trades based on trade_id
+        if trade_id == -1:
```
i really don't like to have to pass an implicit `-1` to the endpoint to get open trades
consider making `trade_id` optional instead - and create an additional endpoint with `/trades/open/custom-data` or similar. 

If it's placed before/above the more generic one, it should work just fine.

xmatthias commented:
@arenstar any plans on finishing this?

arenstar commented:
> @arenstar any plans on finishing this?

Yes i do.

arenstar committed:
fix: endpoints

arenstar committed:
fix: key in _rpc_list_custom_data

arenstar commented:
So ive made some updates now :)

arenstar committed:
Merge branch 'develop' into api-server-list-custom-data

arenstar committed:
fixes: ruff

arenstar committed:
fix

arenstar committed:
fix E501

arenstar committed:
fix: formating

froggleston commented:
Please run mypy locally to fix the remaining formatting issues: https://github.com/freqtrade/freqtrade/blob/develop/CONTRIBUTING.md#run-mypy

Axel-CH committed:
Merge branch 'develop' into api-server-list-custom-data

Axel-CH committed:
fix: update _rpc_list_custom_data with proper typing and custom_data collection loop

xmatthias commented on freqtrade/rpc/rpc.py:
```diff
@@ -1130,15 +1131,15 @@ def _rpc_list_custom_data(
+                if data:
+                    custom_data = [data]
+
+            custom_data.extend(trade.get_all_custom_data())
```
@Axel-CH i think you'll want to keep this line in the else - this way you're ignoring the "key" premise by always querying for "everything" and extending `custom_data` with it.

Axel-CH committed:
fix: update _rpc_list_custom_data to add all custom data only if key is not provided

Axel-CH commented:
@xmatthias do you think there is something to add or change to get this feature merged?

xmatthias commented on freqtrade/rpc/api_server/api_v1.py:
```diff
@@ -508,3 +509,19 @@ def sysinfo():
     return rpc.health()
+
+
+@router.get("/trades/open/custom-data", response_model=list[ListCustomData], tags=["info"])
```
Tags for these endpoints shouldn't be "info".

they should (from a location point of view) be moved near the other "/trades/" calls.

xmatthias commented on freqtrade/rpc/rpc.py:
```diff
@@ -1115,19 +1115,33 @@ def _rpc_delete(self, trade_id: int) -> dict[str, str | int]:
+            # Collect custom data
+            if key:
+                data = trade.get_custom_data(key=key)
+                if data:
+                    custom_data = [data]
```
i don't see any indication that passing "key" restricts this method to a single trade.
with the current implementation however - this becomes an (implicit) fact - so if i call `_rpc_list_custom_data(key="hello")` - i get the data for the very last trade (based on no sorting) - while in reality, i asked for info for all trades (if it exists, obviously).

The fact that tests pass right now highlights the fact that this is also missing in test coverage.

xmatthias commented on freqtrade/rpc/api_server/api_v1.py:
```diff
@@ -508,3 +509,19 @@ def sysinfo():
+
+
+@router.get("/trades/{trade_id}/custom-data", response_model=list[ListCustomData], tags=["info"])
+def list_custom_data(trade_id: int, rpc: RPC = Depends(get_rpc)):
```
i think "key" should be part of (both) endpoints.

xmatthias commented on freqtrade/rpc/api_server/api_v1.py:
```diff
@@ -508,3 +509,19 @@ def sysinfo():
+    """
+    Fetch custom data for all open trades.
+    """
+    return rpc._rpc_list_custom_data()
```
I doubt this will scale - if you've got 100ds or 1000nds of trades here - you'll get all entries back, resulting in a call that either takes ages, consumes tons of memory - or (more likely) - both.

As a consequence, this endpoint will probably need some kind of pagination  / limit - probably / most likely similar to the `/trades` endpoint.

Alternatively, a variant that accepts a trade list might also work - with the assumption that you may not want all trades - but just a couple.

Axel-CH replied:
I've added the pagination management, we will see if the variant accepting a list of trades could be useful.

xmatthias reviewed (changes requested):
@Axel-CH tests are missing for sure - from a visual inspection only, i found about 2-3 points that must be changed - i expect that having tests will incrase what we find.

Axel-CH committed:
chore: relocate custom-data endpoints of api server near trade related endpoint, replace info tag by trading

Axel-CH committed:
fix: _rpc_list_custom_data send custom-data for all trades if a key is provided

Axel-CH committed:
fix: on custom-data endpoints key is now an optional parameter

Axel-CH committed:
chore: implement pagination for _rpc_list_custom_data

Axel-CH committed:
feat: implement pagination for open trades custom-data rpc endpoint

Axel-CH committed:
feat: add retrieval_mode in get_custom_data function to chose between value or full custom_data object

Axel-CH committed:
chore: enhance _rpc_list_custom_data error handling, output format and docstring

Axel-CH committed:
feat: update api schema custom data related classes

Axel-CH committed:
chore: update api custom-data related routes with better not found error handling

Axel-CH committed:
chore: update telegram _list_custom_data according to _rpc_list_custom_data output format change

Axel-CH committed:
test: slight change on expected string in test_telegram_list_custom_data

Axel-CH committed:
test: add test_api_custom_data_single_trade to validate api route behaviour

Axel-CH committed:
test: add test_api_custom_data_multiple_open_trades to validate api route behaviour

Axel-CH committed:
Merge branch 'develop' into api-server-list-custom-data

xmatthias commented on ft_client/freqtrade_client/ft_rest_client.py:
```diff
@@ -484,3 +484,17 @@ def health(self):
         """
         return self._get("health")
+
+    def list_custom_data(self, trade_id=None, key=None):
```
key ain't passed to the request?

would it also make sense to implement the /open/custom-data endpoint?
I'd also move this towards the other /trade implementations - that's especially important for this file - as the sequence in the file determines the sequence in the help section.

xmatthias commented on ft_client/freqtrade_client/ft_rest_client.py:
```diff
@@ -484,3 +484,17 @@ def health(self):
+        if trade_id is not None:
+            params["trade_id"] = trade_id
+
+        return self._get("trades/{tradeid}/custom_data", params=params)
```
I'd strongly suggest to manually test this. this way - it's not working (for several reasons i guess)

Axel-CH replied:
This is the last remaining point, I will continue with those manual tests tomorrow.

Axel-CH replied:
Tested today and both command:

`freqtrade-client --config config.json list_custom_data 1`
`freqtrade-client --config config.json list_open_trades_custom_data`

Are working properly

xmatthias commented on freqtrade/rpc/rpc.py:
```diff
@@ -1115,31 +1115,72 @@ def _rpc_delete(self, trade_id: int) -> dict[str, str | int]:
+                for data_entry in custom_data
+            ]
+
+            # Append result for the trade if any custom data was found.
+            if formatted_custom_data:
+                results.append({"trade_id": trade.id, "custom_data": formatted_custom_data})
```
This section seems odd
why would you first try to create an object and then add it?
Can't the "if" be moved to the top of the `formatted_custom_data` section, skipping that section if nothing is found?

xmatthias commented on freqtrade/rpc/rpc.py:
```diff
@@ -1115,31 +1115,72 @@ def _rpc_delete(self, trade_id: int) -> dict[str, str | int]:
+                message_details = f"found for Trade ID: {trade_id}."
+            elif key:
+                message_details = f"with key '{key}' found for any open trades."
+            raise RPCException(f"No custom_data {message_details}")
```
i'd prefer we stick to `custom-data` terminology (with dash, not underscore)
especially given that we use the same in telegram wording (if there's no error, anyway).

xmatthias commented on freqtrade/rpc/rpc.py:
```diff
@@ -1115,31 +1115,72 @@ def _rpc_delete(self, trade_id: int) -> dict[str, str | int]:
+                results.append({"trade_id": trade.id, "custom_data": formatted_custom_data})
+
+        # Handle case when there is no custom data found across trades.
+        if not results:
+            message_details = "found for any open trades."
+            if key and trade_id:
+                message_details = f"with key '{key}' found for Trade ID: {trade_id}."
+            elif trade_id:
+                message_details = f"found for Trade ID: {trade_id}."
+            elif key:
+                message_details = f"with key '{key}' found for any open trades."
```
I'd prefer a slightly simpler / more concise approach to essentially get to the same end-result
```suggestion
            message_details = ""
            if key:
                message_details += f"with key '{key}' "
            message_details += (
                f"found for Trade ID: {trade_id}." if trade_id else "found for any open trades."
            )
```

Axel-CH replied:
Thank you for the suggestion, I applied it while keeping the "if not results:" check, I suppose it was a mistake in your suggestion.

xmatthias commented on freqtrade/rpc/api_server/api_schemas.py:
```diff
@@ -637,3 +637,18 @@ class Health(BaseModel):
     bot_startup_ts: int | None = None
+
+
+class CustomDataEntry(BaseModel):
+    id: int
+    ft_trade_id: int
+    cd_key: str
+    cd_type: str
+    cd_value: Any
+    created_at: datetime
+    updated_at: datetime | None = None
+
+
+class ListCustomData(BaseModel):
+    trade_id: int
+    custom_data: list[CustomDataEntry]
```
I'm not sure i like the current output format.
trade_id is duplicated (at the 1st level and the 2nd) - and the actual data interpretation is left to the user (is it an int, float, string, ...)?

I also don't think we need (or should) to post the internal ID via API - it's kinda pointless if you think about it - as the user can't do anything with this information - if there's modifications to the value necessary, that'll go over trade_id and cd_key.

(e.g.)

``` jsonc
[
    {
        "trade_id": 5,
        "custom_data": [
            {
                "id": 1,
                "ft_trade_id": 5,
                "cd_key": "test_int",
                "cd_type": "int",
                "cd_value": "1",
                "created_at": "2025-03-13T06:11:22.725051",
                "updated_at": null
            }
        ]
    }
]
```

I'd prefer something around the following:
``` jsonc
[
    {
        "trade_id": 5,
        "custom_data": [
            {
                "key": "test_int",
                "type": "int",
                "cd_value": "1",  // Optional - do we need to duplicate this?
                "value": 1,
                "created_at": "2025-03-13T06:11:22.725051",
                "updated_at": null
            }
        ]
    }
]
```

This would also implicitly fix some potential problems with objects - where python's serialization may work differently than the deser from a different language like JS - which can cause ambiguity and/or even sometimes wrong outputs (assuming a UI or something similar).

We should (probably) also reconsider naming of the response - i don't see why "cd_key" (internal terminology) should be used instead of "key". If you look at how a user would use this in a strategy, the current result is a completely different structure than what it used to be.

Axel-CH replied:
I did the requested changes, I kept "cd_value" as technically it is not duplicate data but a different representation.

xmatthias commented on freqtrade/persistence/trade_model.py:
```diff
@@ -1351,13 +1351,20 @@ def set_custom_data(self, key: str, value: Any) -> None:
         CustomDataWrapper.set_custom_data(trade_id=self.id, key=key, value=value)
 
-    def get_custom_data(self, key: str, default: Any = None) -> Any:
+    def get_custom_data(self, key: str, default: Any = None, retrieval_mode: str = "value") -> Any:
```
I'd prefer not to overload this, but instead using a different method name.

This method is part of the strategy user interface - but i see no reason for a user to actually request this from a strategy.

**edit** which is already there in `get_custom_data_entry()` - making this change pointless.

Axel-CH replied:
I was thinking about that but I thought that in some cases having the last updated date could be useful for users to check if a specific custom data is outdated.

xmatthias replied:
well that's a different topic. If you think that - it should be a separate PR explaining the change and reasoning (including updating the documentation).
That should also explain why `get_custom_data_entry()` wouldn't work for a user in that case.

For this PR - it's out of scope. Please revert.

xmatthias reviewed (changes requested):
A few points so far - the output format is pretty ... rough - and from a  user perspective - not really workable.

the client implementation itself is also broken - essentially  making 50% of the "promise" in the PR description something between missing and broken.

Axel-CH committed:
chore: move list_custom_data closer to trades related functions

Axel-CH committed:
feat: set trade_id as required param in list_custom_data, add key as optional

Axel-CH committed:
feat: add list_open_trades_custom_data to ft rest client

Axel-CH committed:
chore: update rest client custom-data related functions inline comment and help content

Axel-CH committed:
chore: small refactor in _rpc_list_custom_data

Axel-CH committed:
chore: wording update in _rpc_list_custom_data

Axel-CH committed:
chore: simplify error handling in _rpc_list_custom_data

Axel-CH committed:
chore: revert unnecessary get_custom_data changes

Axel-CH committed:
test: update test_telegram_list_custom_data wording

Axel-CH committed:
chore: enhance list custom-data output format

Axel-CH committed:
chore: enhance update rpc_list_custom_data output

Axel-CH committed:
chore: update api schema according rpc_list_custom_data output

Axel-CH committed:
test: update custom-data api related tests according rpc_list_custom_data output

Axel-CH committed:
chore: update telegram _list_custom_data

Axel-CH committed:
test: update test_telegram_list_custom_data

Axel-CH committed:
Merge branch 'develop' into api-server-list-custom-data

Axel-CH committed:
fix: rest client custom-data path

xmatthias commented on freqtrade/rpc/telegram.py:
```diff
@@ -1981,16 +1981,18 @@ async def _list_custom_data(self, update: Update, context: CallbackContext) -> N
-                        f"*Value:* `{result['cd_value']}`",
-                        f"*Create Date:* `{format_date(result['created_at'])}`",
-                        f"*Update Date:* `{format_date(result['updated_at'])}`",
+                        f"*Trade ID:* `{trade_id}`",
```
I don't think tradeId should be in the output.

the telegram command is quite explicit - and only allows `/list_custom_data <tradeid> [key]` - if tradeID is no integer, it'll fail - so it unnecessarily extends the message.

xmatthias commented on ft_client/freqtrade_client/ft_rest_client.py:
```diff
@@ -269,6 +269,36 @@ def trades(self, limit=None, offset=None):
+        if key is not None:
+            params["key"] = key
+
+        return self._get("trades/open/custom-data", params=params)
```
`freqtrade-client --config partial_api.json list_custom_data 5` returns the following:

`{"detail": [{"type": "int_parsing", "loc": ["path", "trade_id"], "msg": "Input should be a valid integer, unable to parse string as an integer", "input": "{tradeid}"}]}`

Axel-CH replied:
Yes I'm on it, I will let you know when it is fixed

xmatthias commented on freqtrade/rpc/rpc.py:
```diff
@@ -1115,31 +1115,68 @@ def _rpc_delete(self, trade_id: int) -> dict[str, str | int]:
+        if trade_id is None:
+            # Get all open trades
+            trades = Trade.session.scalars(
+                Trade.get_trades_query([Trade.is_open.is_(True)]).limit(limit).offset(offset)
```
To reliably use limit / offset - you MUST provide default ordering.

without order clause - databases don't guarantee any particular row order.
in most cases - it's difficult to test as it's "mostly" the insert order  - but it CAN change on bigger tables :D

Axel-CH replied:
Yes I plan to use the same ordering logic as _rpc_trade_history

xmatthias reviewed (changes requested):
Unfortunately, the client ain't working just yet.
postman results however are okish now.

Axel-CH committed:
fix: list_custom_data trade id variable value

Axel-CH committed:
chore: remove trade id from telegram response for list_custom_data

Axel-CH committed:
test: update test after trade id removal from telegram response for list_custom_data

Axel-CH committed:
fix: implement ordering for rpc_list_custom_data

Axel-CH commented:
Did the latest fixes, tested freqtrade-client and both newly added commands
**list_custom_data** and **list_open_trades_custom_data**  are working properly.

xmatthias commented on freqtrade/rpc/rpc.py:
```diff
@@ -1115,31 +1115,72 @@ def _rpc_delete(self, trade_id: int) -> dict[str, str | int]:
+        trades: Sequence[Trade]
+        if trade_id is None:
+            # Get all open trades
+            order_by: Any = Trade.close_date.desc()
```
```suggestion
            order_by: Any = Trade.open_date.desc()
```
Open date? `close_date` will be empty for all open trades.
While unlikely (you'd have to have open trades > limit) - a low "limit" number could still potentially break pagination.

Axel-CH committed:
chore: remove cd_value from rpc custom data output

Axel-CH committed:
test: update tests after removal of cd_value field from list custom-data response

Axel-CH committed:
chore: update schema after removal of cd_value field from list custom-data response

Axel-CH committed:
chore: use open_date for _rpc_list_custom_data output ordering

Axel-CH committed:
chore: use Trade id for _rpc_list_custom_data output ordering

xmatthias committed:
chore: simplify import and docstring

xmatthias committed:
test: simplify assert message

xmatthias reviewed (approved):
seems to work now :+1: 

LGTM :+1:
